// graphql-server-koa/dist/koaApollo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_server_core_1 = require("graphql-server-core");
var GraphiQL = require("graphql-server-module-graphiql");
function graphqlKoa(options) {
    if (!options) {
        throw new Error('Apollo Server requires options.');
    }
    if (arguments.length > 1) {
        throw new Error("Apollo Server expects exactly one argument, got " + arguments.length);
    }
    return function (ctx) {
        return graphql_server_core_1.runHttpQuery([ctx], {
            method: ctx.request.method,
            options: options,
            query: ctx.request.method === 'POST' ? ctx.request.body : ctx.request.query,
        }).then(function (gqlResponse) {
            ctx.set('Content-Type', 'application/json');
            ctx.body = gqlResponse;
        }, function (error) {
            if ('HttpQueryError' !== error.name) {
                throw error;
            }
            if (error.headers) {
                Object.keys(error.headers).forEach(function (header) {
                    ctx.set(header, error.headers[header]);
                });
            }
            ctx.status = error.statusCode;
            ctx.body = error.message;
        });
    };
}
exports.graphqlKoa = graphqlKoa;
function graphiqlKoa(options) {
    return function (ctx) {
        var q = ctx.request.query || {};
        var query = q.query || '';
        var operationName = q.operationName || '';
        var graphiQLString = GraphiQL.renderGraphiQL({
            endpointURL: options.endpointURL,
            subscriptionsEndpoint: options.subscriptionsEndpoint,
            query: query || options.query,
            variables: q.variables && JSON.parse(q.variables) || options.variables,
            operationName: operationName || options.operationName,
            passHeader: options.passHeader,
        });
        ctx.set('Content-Type', 'text/html');
        ctx.body = graphiQLString;
    };
}
exports.graphiqlKoa = graphiqlKoa;
